### PR TITLE
[JENKINS-30002] Improve Util.isOverridden

### PR DESCRIPTION
**Pull request by amuniz**
[JENKINS-30002](https://issues.jenkins-ci.org/browse/JENKINS-30002)

@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/Util.java [hunk 1]
```diff
@@ -33,6 +34,7 @@
 import hudson.util.QuotedStringTokenizer;
 import hudson.util.VariableResolver;
 import hudson.util.jna.WinIOException;
+
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Util.java]: please avoid such unrelated changes

File: core/src/main/java/hudson/Util.java [hunk 2]
```diff
@@ -1415,18 +1422,36 @@ public static Number tryParseNumber(@CheckForNull String numberStr, @CheckForNul
...
         try {
-            return !base.getMethod(methodName, types).equals(
-                    derived.getMethod(methodName,types));
+            return !getMethod(base, methodName, types).equals(getMethod(derived, methodName, types));
         } catch (NoSuchMethodException e) {
             throw new AssertionError(e);
         }
     }
 
+    private static Method getMethod(@Nonnull Class clazz, @Nonnull String methodName, @Nonnull Class... types) throws NoSuchMethodException {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Util.java]: Seems it can return null if an exception happens and superclass is null. The method should be marked as CheckForNull

**Comment by oleg-nenashev** [comment]
Sorry, I'm putting a :bug: on it, because I need more clarification.

It smells that the code of <code>getMethod()</code> breaks the encapsulation. Seems it may return a private method of the superclass with the requested signature, which is wrong even if does not impact <code>isOverriden()</code>. Could you clarify it? Probably, it would be enough to rename a method to a less generic one (or write Javadocs)

File: core/src/main/java/hudson/Util.java (hunk 2 above)
Comment (amuniz) [inline core/src/main/java/hudson/Util.java]: It can not return null, see the last if block.

**Comment by amuniz** [comment]
@oleg-nenashev you are right, it breaks encapsulation, but it's exactly what we want in this case, that's why we are avoiding to use the original ```Class.getMethod``` which only search for ```public``` overrides. 

See the test code for a clear example. Current ```isOverridden``` does not work properly for ```protected``` overridden methods.

(New lines in imports section removed).

**Comment by amuniz** [comment]
@oleg-nenashev BTW did you read the JIRA issue? All this is already explained there.

**Comment by olivergondza** [comment]
I agree with Oleg that this should work for public, protected and package protected methods, though not for private. It is not overriding be definition and it makes no sense for any use case I can imagine.

File: core/src/main/java/hudson/Util.java (hunk 1 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Util.java]: Thanks for fixing it!

File: core/src/main/java/hudson/Util.java (hunk 2 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Util.java]: Sorry, I've missed the obvious code block :(
From my POW, it would be better not to throw runtime exceptions. Null might be better here, because the calling method should be able the case. 

Currently we have a runtime exception, which may be thrown away from the new isOverriden() method and which has no documentation or such runtime exception. This behavior is very hazardous for a widely used method.

**Comment by tfennelly** [comment]
@amuniz yeah, not sure I get why we'd include `private` methods? I think the PR would probably be fine if it included a `!Modifier.isPrivate(method.getModifiers())` check to exclude private methods.

**Comment by amuniz** [comment]
Ok, it makes sense.
Fixed.

**Comment by oleg-nenashev** [comment]
@amuniz 
And WDYT about the runtime exception?

**Comment by amuniz** [comment]
@oleg-nenashev I'm keeping the original behavior. Perhaps there is some code out there relying on this, so I'd prefer to keep it as is. Don't you agree?

**Comment by oleg-nenashev** [comment]
@amuniz 
Yes, I agree. It would be great to just add the <code>AssertionError</code> to Javadoc
Should we also add a check for statics? They are not overridable, but I don't know how this reflection magic works for them.

**Comment by jglick** [comment]
I think it would suffice to throw an `IllegalArgumentException` if the method is not either `public` or `protected`, or if it is `static` or `final`, or if the “derived” class is not assignable to the parent class.

**Comment by amuniz** [comment]
@jglick I think that would break backward compatibility, If someone is catching ```AssertionError``` for some reason, that catch will never be reached anymore.

**Comment by jglick** [comment]
Should also check the comment in `org.jvnet.tiger_types.Types.isOverriding` (no clue where source is—last seen on java.net) regarding overrides of generic methods:

```java
public abstract class Base<T>
    public void setX(T t) {}
    public T getX() {return null;}
}
public class Derived extends Base<Integer> {
    public void setX(Integer i) {}
    public Integer getX() {return 0;}
}
```

Here both methods are overridden in `Derived`. I _think_ the current implementation handles that, but it would be worth testing.

**Comment by jglick** [comment]
> If someone is catching `AssertionError` for some reason…

…then they should be punished!

**Comment by amuniz** [comment]
All fixed.

File: core/src/main/java/hudson/Util.java [hunk 3]
```diff
@@ -1415,16 +1423,36 @@ public static Number tryParseNumber(@CheckForNull String numberStr, @CheckForNul
     }
 
     /**
-     * Checks if the public method defined on the base type with the given arguments
-     * are overridden in the given derived type.
+     * Checks if the method defined on the base type with the given arguments
+     * is overridden in the given derived type.
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Util.java]: I still think <code>@throws</code> would be useful here

**Comment by oleg-nenashev** [comment]
:bee: with a minor comment

**Comment by amuniz** [comment]
@reviewbybees done

**Comment by tfennelly** [comment]
:bee: 

Nice job @amuniz

**Comment by amuniz** [comment]
Would be something additional from my side to have this merged?

**Comment by oleg-nenashev** [comment]
I'm OK with the merge, but it would be better to squash PRs. There are people who are against squashing, but as a past maintainer of an internal Jenkins fork I feel strongly about it.

**Comment by tfennelly** [comment]
@oleg-nenashev yeah, I agree with you on squashing. The vast majority of the history in a PR is just irritating noise; adding annotations, NPE checks and other such.

The merge commit argument (etc) is fine, but I think the PR owner should be allowed to squash the PR before final commit if they want. In fact, is this even a "rule" that the community has agreed upon?

On other OSS projects I have contributed to, you'd get your balls handed to you in a jar if you pushed a multi-commit PR upstream. In fact, I know some that hate merge commits i.e. they want the upstream commit history to read like a book as much as possible.

**Comment by amuniz** [comment]
@oleg-nenashev @tfennelly squashed.

**Comment by oleg-nenashev** [comment]
Thanks @amuniz 

@tfennelly Yes, my position is same. Probably I'll start a thread on jenkinsci-dev and propose such rule.

**Comment by tfennelly** [comment]
@oleg-nenashev or propose/confirm that there is no such rule i.e. it's the PR owner choice.

**Comment by jglick** [comment]
Please, please do not squash commits _inside_ a PR. If you insist on producing a squashed commit for the `master` branch to close the PR, that is one thing (which I disagree with), but it does not require destroying history in the PR.

**Comment by KostyaSha** [comment]
@jglick please please don't merge master into PR before merging it to master.

**Comment by oleg-nenashev** [comment]
@jglick @tfennelly 
I agree that we don't want to request squashing from contributors to save their time. BTW merging of unsquashed commits to the master of a big OSS project is a really bad thing IMO. I'm not going to do it.

BTW I've switched to [this script](https://gist.github.com/oleg-nenashev/ec7dea01dcf272e9b891), which performs the things semi-automatically. So no I can squash commits from others w/o significant efforts.

**Comment by stephenc** [comment]
and @oleg-nenashev forced his personal opinion on squashing commits when merging to master
sad panda

**Comment by jglick** [comment]
> I agree that we don't want to request squashing from contributors to save their time.

No, my concern was rather about destroying the history of activity recorded in the PR. A script like yours at least leaves the PR intact, so we have the information available so long as we host on GitHub, even if information is destroyed in the official project history.

> don't merge master into PR before merging it to master

No idea what that refers to.

**Comment by stephenc** [comment]
>> don't merge master into PR before merging it to master

> No idea what that refers to.

The only guess I have is 955eb7274a02e3549493cd2e1e0de785a0893784 but that wasn't you anyway!

**Comment by oleg-nenashev** [comment]
@stephenc 

> and @oleg-nenashev forced his personal opinion on squashing commits when merging to master

I didn't force anything. My statement was:
> I'm OK with the merge, but it would be better to squash PRs. There are people who are against squashing, but as a past maintainer of an internal Jenkins fork I feel strongly about it.

I just provided my opinion. I didn't block the PR by -1 or bug, so anybody else was able to merge the PR if he thinks differently. It's **my right** not to merge PRs when I'm not 100% happy, but I don't require other contributors or maintainers to follow the squash rule outside plugins I maintain.

**Comment by oleg-nenashev** [comment]
@jglick 

> No, my concern was rather about destroying the history of activity recorded in the PR. A script like yours at least leaves the PR intact, so we have the information available so long as we host on GitHub, even if information is destroyed in the official project history.

I'm not against multi-commit changes when these commits are really reasonable on the atomic level (E.g. "Added new API method" + "Added unit tests for new API" + "Migrated stuff to the new API"). I'm against dozens of commits happening during the PR polishing: typo fixes, minor regressions, etc. I doubt this info will be ever required, but it makes the analysis very difficult when you want to investigate the issue without checking out the entire repo.

Probably keeping full histories in PRs is a convenient approach.
